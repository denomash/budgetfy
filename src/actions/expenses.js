import database from '../firebase/firebase';

// Add expense action
export const addExpense = expense => ({
  type: 'ADD_EXPENSE',
  expense
});

export const startAddExpense = (expenseData = {}) => (dispatch, getState) => {
  const uid = getState().auth.uid;
  const {
    description = '',
    note = '',
    amount = 0,
    createdAt = 0
  } = expenseData;

  const expense = { description, note, amount, createdAt };

  return database
    .ref(`users/${uid}/expenses`)
    .push(expense)
    .then(ref => {
      dispatch(
        addExpense({
          id: ref.key,
          ...expense
        })
      );
    });
};

// Add expense action
export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

export const startRemoveExpense = ({ id }) => (dispatch, getState) => {
  const uid = getState().auth.uid;

  return database
    .ref(`users/${uid}/expenses/${id}`)
    .remove()
    .then(() => {
      dispatch(removeExpense({ id }));
    })
    .catch(err => {
      console.log('Error removing an expense ', err);
    });
};

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

export const startEditExpense = (id, updates) => (dispatch, getState) => {
  const uid = getState().auth.uid;

  return database
    .ref(`users/${uid}/expenses/${id}`)
    .update(updates)
    .then(() => {
      dispatch(editExpense(id, updates));
    });
};

// SET EXPENSES
export const setExpenses = expenses => ({
  type: 'SET_EXPENSES',
  expenses
});

export const startSetExpense = () => (dispatch, getState) => {
  const uid = getState().auth.uid;

  return database
    .ref(`users/${uid}/expenses`)
    .once('value')
    .then(snapshotVal => {
      const expenses = [];
      snapshotVal.forEach(snapshot => {
        expenses.push({
          id: snapshot.key,
          ...snapshot.val()
        });
      });

      dispatch(setExpenses(expenses));
    });
};
