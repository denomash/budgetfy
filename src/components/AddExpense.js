import React from 'react';

const Dashboard = () => {
  return <div>This is the add expense component</div>;
};

export default Dashboard;
