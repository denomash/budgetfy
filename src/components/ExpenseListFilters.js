import React, { Component } from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import {
  setTextFilter,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate
} from '../actions/filters';

export class ExpenseListFilters extends Component {
  state = {
    calendarFocused: null
  };

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = calendarFocused => {
    this.setState({ calendarFocused });
  };

  onTextChange = e => {
    this.props.setTextFilter(e.target.value);
  };

  onSortChange = e => {
    if (e.target.value === 'date') {
      this.props.sortByDate();
    }
    if (e.target.value === 'amount') {
      this.props.sortByAmount();
    }
  };

  render() {
    const { filters } = this.props;
    return (
      <div className="content-container">
        <div className="input-group">
          <div className="input-group__item">
            <input
              type="text"
              value={filters.text}
              onChange={this.onTextChange}
            />
          </div>

          <div className="input-group__item">
            <select value={filters.sortBy} onChange={this.onSortChange}>
              <option value="date">Date</option>
              <option value="amount">Amount</option>
            </select>
          </div>

          <div className="input-group__item">
            <DateRangePicker
              startDate={filters.startDate}
              endDate={filters.endDate}
              onDatesChange={this.onDatesChange}
              focusedInput={this.state.calendarFocused}
              onFocusChange={this.onFocusChange}
              showClearDates={true}
              numberOfMonths={1}
              isOutsideRange={() => false}
            />
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => {
  return {
    filters: state.filters
  };
};

const mapDispatchToProps = dispatch => {
  return {
    setStartDate: startDate => dispatch(setStartDate(startDate)),
    setEndDate: endDate => dispatch(setEndDate(endDate)),
    setTextFilter: text => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount())
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ExpenseListFilters);
