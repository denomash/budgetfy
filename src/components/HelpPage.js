import React from 'react';

const HelpPage = () => {
  return <div>TThis is the Help component</div>;
};

export default HelpPage;
