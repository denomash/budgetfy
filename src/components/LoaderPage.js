import React from 'react';

const LoaderPage = () => {
  return (
    <div className="loader">
      <img className="loader__image" src="/images/loader.gif" />
    </div>
  );
};

export default LoaderPage;
