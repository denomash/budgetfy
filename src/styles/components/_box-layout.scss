.box-layout {
  display: flex;
  align-items: center;
  justify-content: center;
  background: linear-gradient(
      rgba(255, 255, 255, 0.2),
      rgba(255, 255, 255, 0.2)
    ),
    url('/images/new york.jpg');
  background-size: cover;
  height: 100vh;
  width: 100vw;

  &__box {
    background: fade-out(#fff, 0.1);
    border-radius: 3px;
    box-shadow: 0 3px 6px rgba(45, 45, 45, 0.2);
    text-align: center;
    width: 62rem;
    padding: $l-size $m-size;
  }

  &__title {
    margin: 0 0 2.6rem 0;
    line-height: 1;
  }

  &__center {
    display: flex;
    justify-content: center;
    align-items: center;
  }

  &__content {
    display: flex;
    justify-content: space-between;
    align-items: center;
    width: 90%;
  }

  &__divider {
    width: 2px;
    background: #b9b9b9;
    height: 170px;
  }
}
