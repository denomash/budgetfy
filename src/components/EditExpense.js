import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';

export class EditExpense extends React.Component {
  constructor(props) {
    super(props);
  }
  onEditExpense = expense => {
    this.props.editExpense(expense);
    this.props.history.push('/');
  };

  onRemoveExpense = () => {
    this.props.removeExpense({ id: this.props.expense.id });
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <h1>Edit Expense</h1>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.onEditExpense}
        />

        <button onClick={this.onRemoveExpense}>Remove</button>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(
      expense => expense.id === props.match.params.id
    )
  };
};

const mapDispatchToProps = (dispatch, props) => {
  return {
    addExpense: expense => dispatch(editExpense(props.expense.id, expense)),
    removeExpense: id => dispatch(removeExpense(id))
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditExpense);
