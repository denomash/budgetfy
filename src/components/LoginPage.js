import React from 'react';
import { connect } from 'react-redux';

import { startLoggin } from '../actions/auth';

export const LoginPage = ({ startLoggin }) => {
  return (
    <div>
      <button onClick={startLoggin}>Login</button>
    </div>
  );
};

const mapDispatchToProps = dispatch => ({
  startLoggin: () => dispatch(startLoggin())
});

export default connect(
  null,
  mapDispatchToProps
)(LoginPage);
