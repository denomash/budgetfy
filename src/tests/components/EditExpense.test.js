import React from 'react';
import { shallow } from 'enzyme';
import { EditExpense } from '../../components/EditExpense';
import expenses from '../fixtures/expenses';

let editExpense, startRemoveExpense, history, wrapper;

beforeAll(() => {
  editExpense = jest.fn();
  startRemoveExpense = jest.fn();
  history = { push: jest.fn() };

  wrapper = shallow(
    <EditExpense
      editExpense={editExpense}
      startRemoveExpense={startRemoveExpense}
      history={history}
      expense={expenses[1]}
    />
  );
});

test('Should display edit expense correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('Should handle edit expense', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]);

  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(editExpense).toHaveBeenLastCalledWith(expenses[1].id, expenses[1]);
});

test('Should handle remove expense', () => {
  wrapper.find('button').simulate('click');

  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(startRemoveExpense).toHaveBeenLastCalledWith({ id: expenses[1].id });
});
