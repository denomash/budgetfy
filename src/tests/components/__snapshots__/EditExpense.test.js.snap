// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should display edit expense correctly 1`] = `
<div>
  <h1>
    Edit Expense
  </h1>
  <ExpenseForm
    expense={
      Object {
        "amount": 98000,
        "createdAt": -864000000,
        "description": "Electricity",
        "id": "2",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
